---
title: "Characterization of SV 577_579"
author: "Daoming Wang, Lei Liu"
date: "2023/04/26"
output:
  html_document: 
    theme: flatly
    highlight: espresso
    toc: true
    toc_depth: 4
    toc_float: true
  word_document: default
  pdf_document:
    includes:
      in_header: header.tex
      keep_tex: yes
      latex_engine: xelatex
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1 Preparation

### 1.1 Import

```{r 1.1, echo=TRUE, message=FALSE, warning=FALSE, paged.print=FALSE}
suppressMessages(source("functions.R"))
```

### 1.2 Inputs

```{r 1.2, echo=TRUE}
emale_genes <- read.table("00.rawData/genome/genomes.txt", header=TRUE,stringsAsFactors = F,sep="\t")
emale_seqs <- read.table("00.rawData/genome/genomesSeq.txt", header=TRUE,stringsAsFactors = F,sep="\t")
emale_ava <- read.table("00.rawData/genome/genomesLink.txt", header=T ,stringsAsFactors = F,sep="\t")
emale_genes$name <- factor(emale_genes$name, levels =c("dinB","lacC","ptsH","gatZ-kbaZ","nagA","agaS","agaF","gatY-kbaY","GH109","rhaR","agaV","agaC","agaD","immR"))

tree <- read.tree("00.rawData/tree/ATCC.nwk")
strain_info <- read.delim('00.rawData/tree/group.txt', row.names = 1, sep = '\t')
example_genes <- read.table("00.rawData/tree/example_genes_577_579_direct.txt",fill=TRUE, header=TRUE,sep="\t")

tree_real   <- read.tree("00.rawData/tree/concatenated.nwk")
tree_galnac <- read.tree("00.rawData/tree/RAxML_bestTree.GalNAc.nwk")
```

## 2 Tree and gene organization of all Fprau strains

```{r 2}
group <- split(row.names(strain_info), strain_info$GalNAc_utilization)
SV_status <- split(row.names(strain_info), strain_info$SV_status)

emale_genes$name[is.na(emale_genes$name)] <- ""
emale_genes$name <- factor(emale_genes$name, levels = c("dinB","lacC","ptsH","gatZ-kbaZ","nagA","agaS","agaF","gatY-kbaY","GH109","rhaR","agaV","agaC","agaD","immR", ""))

example_genes.color <- c("grey50", wes_palette("Zissou1", length(levels(emale_genes$name))-2, type = "continuous"),"grey80") 

p_genome <- gggenomes(genes=emale_genes, seqs=emale_seqs, links=emale_ava,adjacent_only = FALSE)
p_genes <- p_genome + 
  geom_bin_label()+
  geom_seq() +  
  geom_gene(aes(fill=name, color = name)) +
  geom_link(aes(fill=name, color = name))+
  geom_gene(aes(fill=name, color = name))+
  scale_fill_manual("Genes", 
                    breaks = levels(emale_genes$name),
                    values = example_genes.color)+
  scale_color_manual("Genes", 
                     breaks = levels(emale_genes$name),
                     values = example_genes.color)+
  geom_gene_tag(aes(label=name), nudge_y=0.1, check_overlap = TRUE)+
  theme(legend.key.size = unit(0.5, 'cm'),
        legend.key.height = unit(0.5, 'cm'), 
        legend.key.width = unit(0.5, 'cm'), 
        legend.title = element_text(size=8), 
        legend.text = element_text(size=6),
        plot.margin = margin(2, 2, 2, 2))

p_tree <- ggtree(groupOTU(tree,group), branch.length='none') + 
  geom_tiplab(aes(label = paste0("italic('  ", label, "')")), parse = TRUE, size = 3.5) +
  geom_tippoint(aes(color = group),size = 2.5, show.legend = T) + 
  xlim_tree(9) +
  scale_x_continuous(expand=c(0,4)) +
  scale_color_manual("GalNAc utilization", values =  rev(c('orange','gray')))+
  theme(strip.text=element_blank(),
        panel.spacing=unit(0, 'cm'),
        legend.position="right")

# Main Fig. 3c
ggsave("04.GalNAc_SV/snpPhyTree_synteny.pdf",
       p_tree +p_genes+ plot_layout(widths = c(3,4)),
       width=12, height=4.5)
```

## 3 Compare trees

Infered the phylogenetic trees based on (1) phylogeny-specific marker genes and (2) GalNAc metabolism genes, then compare two trees.

```{r 3}
tree_real$tip.label<-tree_real$tip.label %>% str_replace_all("\'", "")

p1<-ggtree(tree_real, layout='roundrect', branch.length='none')
p2<-ggtree(tree_galnac, layout='roundrect', branch.length='none') 

d1 <- p1$data
d2 <- p2$data
d2$x <- max(d2$x) - d2$x + max(d1$x) + 10
dd <- bind_rows(d1, d2) %>% 
  filter(!is.na(label))

## reverse x-axis and 
## set offset to make the tree on the right-hand side of the first tree

pdf("04.GalNAc_SV/compare_tree.pdf", width = 6, height = 4) # Supplementary Figure 6a
p1 + geom_tree(data=d2, layout='roundrect')+
  geom_line(aes(x, y, group=label), data=dd %>% dplyr::filter(isTip==TRUE), color='grey')+
  geom_tiplab(geom = 'text', colour = "blue") +
  geom_tiplab(geom = 'text',data = d2, hjust=1,  colour = "firebrick")
dev.off()

# Tree comparison: dissimilarity between two trees
tree.distance <- TreeDistance(tree_real, tree_galnac) # 0.7324924

# between-strain distance comparison
dist_real <- cophenetic(tree_real)
dist_galnac <- cophenetic(tree_galnac)
dist_galnac <- dist_galnac[rownames(dist_real), colnames(dist_real)]

dist_galnac.df <- mat2df(dist_galnac)
dist_real.df <- mat2df(dist_real)

dist.df <- data.frame(Strain1 = dist_real.df$id1,
                      Strain2 = dist_real.df$id2,
                      Real_dist = dist_real.df$value,
                      GalNAc_dist = dist_galnac.df$value)

pdf("04.GalNAc_SV/compare_distance.scatter.pdf", width = 4, height = 4) # Supplementary Figure 6b
ggplot(dist.df, aes(Real_dist, GalNAc_dist))+
  geom_point(alpha = 0.7)+
  xlab("Phylogenetic distance based on marker genes")+
  ylab("Phylogenetic distance based on GalNAc genes") +
  theme_linedraw()+
  theme(legend.position = "bottom",
        panel.grid.minor = element_line(colour  = "grey80"))
dev.off()


dist.df.long <- rbind(dist_real.df, dist_galnac.df)
dist.df.long$Distance <- c(rep("Real_dist",   nrow(dist_real.df)),
                           rep("GalNAc_dist", nrow(dist_galnac.df)))
dist.df.long$Pair <- c(1:nrow(dist_real.df), 1:nrow(dist_galnac.df))

pdf("04.GalNAc_SV/compare_distance.boxplot.pdf", width = 2.5, height = 4) # Supplementary Figure 6c
ggplot(dist.df.long, aes(Distance, value))+
  geom_boxplot(aes(color = Distance))+
  geom_point(aes(color = Distance),alpha = 0.7)+
  geom_line(aes(group = Pair),color = "grey20",alpha = 0.7)+
  xlab("Calculation approach")+
  ylab("Phylogenetic distance") +
  theme_linedraw()+
  theme(legend.position = "none",
        panel.grid.minor = element_line(colour  = "grey80"))
dev.off()


```
